clang/docs/HLSL/RootSignature.rst

```diff
@@ -0,0 +1,257 @@
+                      'COMPARISON_EQUAL' | 'COMPARISON_LESS_EQUAL' | 
+                      'COMPARISON_GREATER' | 'COMPARISON_NOT_EQUAL' | 
+                      'COMPARISON_GREATER_EQUAL' | 'COMPARISON_ALWAYS'
+
+    STATIC_BORDER_COLOR : 'STATIC_BORDER_COLOR_TRANSPARENT_BLACK' | 
+                          'STATIC_BORDER_COLOR_OPAQUE_BLACK' | 
+                          'STATIC_BORDER_COLOR_OPAQUE_WHITE'
+
+
+Serialized format
+======================
+The root signature string is parsed and serialized into a binary format. The
+binary format is a sequence of bytes that can be used to create a root signature
+object in the Direct3D 12 API. The binary format is defined by the
+`D3D12_ROOT_SIGNATURE_DESC (for rootsig_1_0)
+<https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_root_signature_desc>`_
+or `D3D12_ROOT_SIGNATURE_DESC1 (for rootsig_1_1)
+<https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_root_signature_desc1>`_ 
+structure in the Direct3D 12 API. (With the pointers translated to offsets.)
+
+TODO: replace the following link with actual serialized format. 
+    DXC implementation of the serialization could be find `here
+    <https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/DxilRootSignature/DxilRootSignatureSerializer.cpp#L41>`_
+
+Implementation Details
+======================
+
+The root signature string will be parsed in Clang. 
+The parsing 
+will happened when build HLSLRootSignatureAttr or when build standalone root 
+signature blob. 
+
+The root signature parsing will generate a HLSLRootSignatureAttr with member 
+represents the root signature string and the parsed information for each 
+resource in the root signature. It will bind to the entry function in the AST. 
+
+For case compile to a standalone root signature blob, the HLSLRootSignatureAttr 
+will be bind to the TU.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated 
+into a global variable with struct type to express the layout and a constant 
+initializer to save things like space and NumDescriptors in LLVM IR. 
+
+CGHLSLRuntime will generate metadata to link the global variable as root 
+signature for given entry function or just nullptr for the standalone root 
+signature blob case. 
+
+In LLVM DirectX backend, the global variable will be serialized and saved as 
+another global variable with section 'RTS0' with the serialized root signature 
+as initializer. 
+The MC ObjectWriter for DXContainer will take the global and write it to the 
+correct part based on the section name given to the global.
+
+In DXIL validation, the root signature part will be deserialized and check if 
+resource used in the shader (the information is in pipeline state validation 
+part) exists in the root signature. 
+
+Same check could be done in Sema as well. But at AST level, it is impossible 
+to identify unused resource which will be removed later. And the resource 
```

Comment:
Don't we have a rewriter that removes unused globals? That makes me think it is at least possible to handle some of these cases in Sema. I think we need some more thought here and some specifics about what cases we can handle in Sema and what cases we need to handle later.

For upstream these shouldn't be handled in DXIL validation, but rather we should treat these as fatal errors from the backend, so we need to know these cases to ensure that the upstream errors are correctly implemented and tested.